shared-bindings/ps2io/Ps2.c

```diff
@@ -0,0 +1,194 @@
+//|
+//|     import ps2io
+//|     import board
+//|
+//|     kbd = ps2io.Ps2(board.D10, board.D11)
+//|
+//|     while len(kbd) == 0:
+//|         pass
+//|     print(kbd.get_byte())
+//|     print(kbd.send_byte(0xed))
+//|     print(kbd.send_byte(0x02))
+//|
+STATIC mp_obj_t ps2io_ps2_make_new(const mp_obj_type_t *type, size_t n_args, const mp_obj_t *pos_args, mp_map_t *kw_args) {
+    enum { ARG_datapin, ARG_clkpin };
+    static const mp_arg_t allowed_args[] = {
+        { MP_QSTR_datapin, MP_ARG_REQUIRED | MP_ARG_OBJ },
+        { MP_QSTR_clkpin, MP_ARG_REQUIRED | MP_ARG_OBJ },
+    };
+    mp_arg_val_t args[MP_ARRAY_SIZE(allowed_args)];
+    mp_arg_parse_all(n_args, pos_args, kw_args, MP_ARRAY_SIZE(allowed_args), allowed_args, args);
+    assert_pin(args[ARG_clkpin].u_obj, false);
+    assert_pin(args[ARG_datapin].u_obj, false);
+    const mcu_pin_obj_t* clkpin = MP_OBJ_TO_PTR(args[ARG_clkpin].u_obj);
+    assert_pin_free(clkpin);
+    const mcu_pin_obj_t* datapin = MP_OBJ_TO_PTR(args[ARG_datapin].u_obj);
+    assert_pin_free(datapin);
+
+    ps2io_ps2_obj_t *self = m_new_obj(ps2io_ps2_obj_t);
+    self->base.type = &ps2io_ps2_type;
+
+    common_hal_ps2io_ps2_construct(self, datapin, clkpin);
+
+    return MP_OBJ_FROM_PTR(self);
+}
+
+//|   .. method:: deinit()
+//|
+//|      Deinitialises the Ps2 and releases any hardware resources for reuse.
+//|
+STATIC mp_obj_t ps2io_ps2_deinit(mp_obj_t self_in) {
+    ps2io_ps2_obj_t *self = MP_OBJ_TO_PTR(self_in);
+    common_hal_ps2io_ps2_deinit(self);
+    return mp_const_none;
+}
+STATIC MP_DEFINE_CONST_FUN_OBJ_1(ps2io_ps2_deinit_obj, ps2io_ps2_deinit);
+
+//|   .. method:: __enter__()
+//|
+//|      No-op used by Context Managers.
+//|
+//  Provided by context manager helper.
+
+//|   .. method:: __exit__()
+//|
+//|      Automatically deinitializes the hardware when exiting a context. See
+//|      :ref:`lifetime-and-contextmanagers` for more info.
+//|
+STATIC mp_obj_t ps2io_ps2_obj___exit__(size_t n_args, const mp_obj_t *args) {
+    (void)n_args;
+    common_hal_ps2io_ps2_deinit(args[0]);
+    return mp_const_none;
+}
+STATIC MP_DEFINE_CONST_FUN_OBJ_VAR_BETWEEN(ps2io_ps2___exit___obj, 4, 4, ps2io_ps2_obj___exit__);
+
+//|   .. method:: get_byte()
+//|
+//|     Removes and returns the oldest received byte. When buffer
+//|     is empty, returns a negative number.
+//|
+STATIC mp_obj_t ps2io_ps2_obj_get_byte(mp_obj_t self_in) {
+    ps2io_ps2_obj_t *self = MP_OBJ_TO_PTR(self_in);
+    raise_error_if_deinited(common_hal_ps2io_ps2_deinited(self));
+
+    return MP_OBJ_NEW_SMALL_INT(common_hal_ps2io_ps2_get_byte(self));
+}
+MP_DEFINE_CONST_FUN_OBJ_1(ps2io_ps2_get_byte_obj, ps2io_ps2_obj_get_byte);
+
+//|   .. method:: send_byte()
+//|
+//|     Sends a byte to PS/2. Returns the response byte (positive)
+//|     or a communication error code (negative). The response is
+//|     typically an ACK (0xFA). Some commands return additional data
+//|     bytes that should be received via get_byte().
```

Comment:
Would it make more sense to not return anything and have error raise an exception? That way one always gets data from `get_byte`. Exceptions can also have data in them (like OSError has an error number).

Reply:
I'd prefer to return the byte (typically an ACK) because it is kind of out-of band. If you use get_byte()^Wpopleft() to get the response, there could be other bytes there already, and flushing() the buffer before calling send_byte() would not be 100% guaranteed because a byte can arrive between the last popleft() and send_byte(). I agree raising and exception in place of negative returns which current version uses to signal errors.